quinn-proto/src/connection/streams/recv.rs

```diff
@@ -312,56 +274,75 @@ impl<'a> Chunks<'a> {
                         ShouldTransmit(false),
                         true,
                     );
+                    self.state = ChunksState::Finished(ShouldTransmit(true));
                     Ok(None)
                 } else {
+                    let should_transmit =
+                        Self::done(rs, self.read, self.id, self.streams, self.pending);
+                    let old = mem::replace(
+                        &mut self.state,
+                        ChunksState::Error(ReadError::Blocked, should_transmit),
+                    );
+
+                    let rs = match old {
+                        ChunksState::Readable(rs) => rs,
+                        _ => unreachable!(),
+                    };
```

Comment:
Could we move out of `self.state` at the start of the function without running afoul of borrowck, and thereby avoid the `unreachable!()` here?

Reply:
Yup, that looks to make more sense.

Reply:
It does end up being a little ugly, having to restore the state in 3 additional places. See what you think?

Reply:
(Continued in https://github.com/quinn-rs/quinn/pull/1003#discussion_r581647348.)